src/services/refactors/convertToOptionalChainExpression.ts

```diff
@@ -157,31 +159,44 @@ namespace ts.refactor.convertToOptionalChainExpression {
     /**
      * Returns subchain if chain begins with subchain syntactically.
      */
-    function getMatchingStart(chain: Expression, subchain: Expression): PropertyAccessExpression | Identifier | undefined {
-        return (isIdentifier(subchain) || isPropertyAccessExpression(subchain)) &&
-        chainStartsWith(chain, subchain) ? subchain : undefined;
+    function getMatchingStart(chain: Expression, subchain: Expression): PropertyAccessExpression | ElementAccessExpression | Identifier | undefined {
+        return (isIdentifier(subchain) || isPropertyAccessExpression(subchain) || isElementAccessExpression(subchain))
+            && chainStartsWith(chain, subchain) ? subchain : undefined;
     }
 
     /**
      * Returns true if chain begins with subchain syntactically.
      */
     function chainStartsWith(chain: Node, subchain: Node): boolean {
         // skip until we find a matching identifier.
-        while (isCallExpression(chain) || isPropertyAccessExpression(chain)) {
-            const subchainName = isPropertyAccessExpression(subchain) ? subchain.name.getText() : subchain.getText();
-            if (isPropertyAccessExpression(chain) && chain.name.getText() === subchainName) break;
+        while (isCallExpression(chain) || isPropertyAccessExpression(chain) || isElementAccessExpression(chain)) {
+            if (getTextOfChain(chain) === getTextOfChain(subchain)) break;
             chain = chain.expression;
         }
-        // check that the chains match at each access.  Call chains in subchain are not valid.
-        while (isPropertyAccessExpression(chain) && isPropertyAccessExpression(subchain)) {
-            if (chain.name.getText() !== subchain.name.getText()) return false;
+        // check that the chains match at each access. Call chains in subchain are not valid.
+        while ((isPropertyAccessExpression(chain) && isPropertyAccessExpression(subchain)) ||
+               (isElementAccessExpression(chain) && isElementAccessExpression(subchain))) {

```

Comment:
@DanielRosenwasser I used this condition in order to avoid typecasting, because for proposed change need to cast `subchain` to

```ts
(subchain as PropertyAccessExpression | ElementAccessExpression).expression;
```